File: Python/llParser.py
from grammar import *

class LLParser(object):
    def __init__(self, grammar):
        # Maps (lhs, first token) to production
        self.table = {}
        self.firstSets = {}
        self.followSets = {}
        self.derivesEpsilon = {}
        self.grammar = grammar

        # Populates the parse table and first set for the lhs symbol. Also indicates whether the symbol can derive epsilon
        def populateFirstSet(lhs):
            self.firstSets[lhs] = set([])
            emptyProduction = False
            productions = self.grammar.getProductions(lhs)
            for production in productions:
                i = 0
                # For each production run through all its symbols in order
                while i < len(production):
                    frontSymbol = production[i]
                    if isTerminal(frontSymbol):
                        # If the symbol is a terminal, update the first set and parse table with it.
                        self.table[(lhs, frontSymbol)] = production
                        self.firstSets[lhs].add(frontSymbol)
                        # Since the production will add no more to the first set, move on to the next one
                        break
                    else:
                        # If the symbol is non-terminal, find out its first set recursively
                        if frontSymbol not in self.firstSets:
                            populateFirstSet(frontSymbol)
                        # The first set of lhs includes the first set of the symbol, which is also added to the parse table
                        for token in self.firstSets[frontSymbol]:
                            self.firstSets[lhs].add(token)
                            self.table[(lhs, token)] = production
                        # If the current symbol derives epsilon, then the next symbol will contribute to the first set as well
                        # Otherwise we can continue to the next production
                        if not self.derivesEpsilon[frontSymbol]:
                            break
                    i += 1
                # If every symbol in the production derives epsilon, that means this production derives epsilon as well
                if i >= len(production):
                    emptyProduction = True
            # If any production derives epsilon, lhs derives epsilon
            self.derivesEpsilon[lhs] = emptyProduction

        # Populates the follow sets of all non-terminals without recursion. Does not populate parse table
        # def populateFollowSet():
        #     # Initialize sets
        #     for rule in grammar.rules:
        #         self.followSets[rule.lhs] = set([])

        #     changing = True
        #     #Loop the algorithm until no more changes to follow sets occur
        #     while changing:
        #         changing = False
        #         for rule in self.grammar.rules:
        #             # For each production of each rule, compute follow sets for every nonterminal symbol by accumulating a list of symbols to add
        #             for production in rule.productions:
        #                 addlhs = True
        #                 symbolsToAdd = set([])
        #                 # Read the symbols backwards
        #                 for symbol in production[::-1]:
        #                     if isTerminal(symbol):
        #                         # If the symbol is terminal, reset the symbol list and add the terminal
        #                         symbolsToAdd = set([symbol])
        #                         addlhs = False
        #                     else:
        #                         # If the symbol is non-terminal, first check if all the symbols after it derives epsilon
        #                         # If so, include the rule's follow set in the symbol list
        #                         if addlhs:
        #                             symbolsToAdd.update(self.followSets[rule.lhs])
        #                         # Also track if the symbol's follow set changes due to symbol list before adding the symbol list to the symbol's follow set
        #                         if not (symbolsToAdd <= self.followSets[symbol]):
        #                             changing = True
        #                         self.followSets[symbol].update(symbolsToAdd)

        #                         # If the symbol is not empty, reset the symbol list. Always add its first set to the symbols
        #                         if self.derivesEpsilon[symbol]:
        #                             symbolsToAdd.update(self.firstSets[symbol])
        #                         else:
        #                             symbolsToAdd = self.firstSets[symbol]
        #                             addlhs = False


        # Compute all first sets and their corresponding table entries
        for rule in self.grammar.rules:
            populateFirstSet(rule.lhs)

        # Do same thing for all follow sets of nonterminals that derive epsilon
        # populateFollowSet()
        # for rule in self.grammar.rules:
        #     if self.derivesEpsilon[rule.lhs]:
        #         for token in self.followSets[rule.lhs]:
        #             self.table[(rule.lhs, token)] = []

    # Does an LL parse of a token list based on the parse table
    def parse(self, startSymbol, tokens):
        tokens += ['$']
        stack = [startSymbol]
        pos = 0
        # Parse until stack is empty or input has been completely parsed
        while stack != [] and pos < len(tokens):
            symbol = stack.pop()
            # If top of stack is terminal, match the next token to it and advance the input
            if isTerminal(symbol):
                if symbol == tokens[pos]:
                    pos += 1
                else:
                    return False
            # If top of stack is non-terminal, insert the correct production backwards into the stack based on the table
            else:
                if (symbol, tokens[pos]) in self.table:
                    reversedProduction = self.table[(symbol, tokens[pos])][::-1]
                    for s in reversedProduction:
                        stack.append(s)
                elif not self.derivesEpsilon[symbol]:
                    return False
        if pos == len(tokens) and stack == []:
            return True
        else:
            return False


grammar = Grammar(
    """
    Start -> Exp $ ;
    Exp -> Term ExpTail ;
    ExpTail -> + Exp | - Exp | epsilon ;
    Term -> Factor TermTail ;
    TermTail -> * Term | / Term | epsilon ;
    Factor -> Value FactorTail | - Factor;
    FactorTail -> ^ Factor | epsilon ;
    Value -> Num | ( Exp ) ;
    Num -> 6 ;
    """)

parser = LLParser(grammar)
for entry in sorted(parser.table):
    print entry, parser.table[entry]
print ""
tokens = "6 * 6 + 6".split()
print parser.parse("Start", tokens)

File: C++/LLParser.cpp
#include "LLParser.h"

//Populates the LL parse table for a given LHS symbol
void LLParser::populateTable(int lhs, std::unordered_set<int> &firstSet){
    //Only run function once for every symbol. By default symbols do not derive epsilon
    if (derivesEpsilon[lhs] > -2){
        return;
    }
    derivesEpsilon[lhs] = -1;
    //For each production of the given rule (i is start of production)
    for (int i = ruleStart(lhs); i < ruleStart(lhs+1); i = nextProduction(i)){
        std::unordered_set<int> descendantSet;
        int j;
        //For each symbol in the production (j is index of symbol)
        for (j = i+1; j <= i+grammar[i]; j++){
            int frontSymbol = grammar[j];
            //If terminal symbol was found, then update the first set and parse table for this production
            if (isTerminal(frontSymbol)){
                table(lhs, frontSymbol) = i;
                firstSet.insert(frontSymbol);
                //Go to next production
                break;
            }
            //Otherwise recursively populate the symbol's first set as well
            else{
                populateTable(toRuleCount(frontSymbol), descendantSet);
                //If the symbol derives epsilon, the following symbol is part of first set as well
                if (derivesEpsilon[toRuleCount(frontSymbol)] < 0){
                    break;
                }
            }
        }
        //Run through the descendant first sets and add them to the current first set and the parse table for current production
        for (auto p: descendantSet){
            firstSet.insert(p);
            table(lhs, p) = i;
        }
        //If entire production has been considered, it means that every symbol derives epsilon, so the lhs does so as well
        if (j > i+grammar[i]){
            derivesEpsilon[lhs] = i;
        }
    }
}

//Constructor populates LL parse table
LLParser::LLParser(char* grammarConfig, Lexer *lexptr) : BaseParserGenerator(grammarConfig, lexptr){
    //Initialize epsilon array to -2 (unseen)
    for (int i=0; i<toRuleCount(ruleNum); i++){
        derivesEpsilon[i] = -2;
    }
    //Populate the parse table
    for (int i=0; i<toRuleCount(ruleNum); i++){
        populateTable(i, std::unordered_set<int>());
    }
}

LLParser::~LLParser(){
    delete[] derivesEpsilon;
    deleteValues(valueStack.size());
}

//Deletes x number of ParserValues on the value stack. Responsible for cleaning the associated memory
void LLParser::deleteValues(int count){
    for (int i=0; i<count; i++){
        if (valueStack.back().toDelete){
            delete valueStack.back().ptr;
        }
        valueStack.pop_back();
    }
}

//Reset all parse variables and shift until next reduction 
ParseStatus LLParser::parse(char* input){
    curpos = input;
    prevpos = input;
    deleteValues(valueStack.size());
    symbolStack.clear();
    //Initialize the symbolStack with starting symbol
    symbolStack.push_back(toRuleNum(0));
    lexptr->reset();
    curTokenNum = next();
    return shiftHelper();
}

//Complete current reduction by replacing the rhs values on the value stack with input lhs value.
//Checks for end of parse then continue shifting.
ParseStatus LLParser::reduce(void *lhsValue, bool toDelete){
    //If the parse stack is empty AND the string has been fully parsed, the parse is done
    if (symbolStack.empty() && curTokenNum==0){
        return DONE;
    }
    //If only the parse stack is empty, then the expected token should be \0, since the parse expects end of input
    else if (symbolStack.empty()){
        expectedSymbol = 0;
        return SYNTAXERROR;
    }
    //Pop off reduction token
    symbolStack.pop_back();
    //Replace appropriate number of values off the value stack with user inputted value
    deleteValues(curSymbolCount);
    ParseValue value;
    value.ptr = lhsValue;
    value.toDelete = toDelete;
    valueStack.push_back(value);
    //Shift until next reduction
    return shiftHelper();
}

//Return unincremented lhs symbol number
int LLParser::lhsNum(){
    return toRuleCount(curLhs);
}
//Return production number
int LLParser::prodNum(){
    return curProdNum;
}
//Return pointer to value of nth reduced symbol
void *LLParser::rhsVal(int pos){
    //std::cout << curSymbolCount;
    return valueStack[valueStack.size() - curSymbolCount + pos].ptr;
}

int LLParser::curToken(){
    return curTokenNum;
}

//Builds a list of expected tokens/chars and returns them
std::vector<int> LLParser::expectedTokens(){
    std::vector<int> expected;
    //If the expected symbol is a terminal, then that symbol is the only expected token
    if (isTerminal(expectedSymbol)){
        expected.push_back(expectedSymbol);
        return expected;
    }
    //If the symbol is nonterminal, then the expected tokens are all that qualify as lookahead for that symbol
    //Search the parse table for the tokens
    for (int i=0; i<tokenNum; i++){
        if (table(toRuleCount(expectedSymbol), i) >= 0){
            expected.push_back(i);
        }
    }
    return expected;
}

//Shift and expand tokens onto the parse stack until the next reduction happens
//Reductions are represented with a -ve value on the stack, representing the position of the production in the grammar
ParseStatus LLParser::shiftHelper(){
    //If the parse stack is empty AND the string has been fully parsed, the parse is done
    if (symbolStack.empty() && curTokenNum==0){
        return DONE;
    }
    //If only the parse stack is empty, then the expected token should be \0, since the parse expects end of input
    else if (symbolStack.empty()){
        expectedSymbol = 0;
        return SYNTAXERROR;
    }
    //Stop at next reduction
    while (symbolStack.back() > 0){
        //Pop symbol off parse stack
        int symbol = symbolStack.back();
        symbolStack.pop_back();
        //If the symbol is a terminal, try to match it to the current token. Get the next token
        if (isTerminal(symbol)){
            if (symbol == curTokenNum){
                //Push the shifted token string onto the value stack
                addTokenValue();
                curTokenNum = next();
            }
            else{
                expectedSymbol = symbol;
                return SYNTAXERROR;
            }
        }
        //If symbol is nonterminal, insert the reduction symbol and the correct production backwards into the parse stack based on parse table
        else{
            int productionPos = table(toRuleCount(symbol), curTokenNum);
            //If table entry doesn't exist for current lhs and token, then see if the lhs derives epsilon
            if (productionPos < 0){
                if (derivesEpsilon[toRuleCount(symbol)] > -1){
                    //Use the epsilon-deriving production as the reducing production
                    productionPos = derivesEpsilon[toRuleCount(symbol)];
                }
                //If lhs doesn't derive epsilon, then syntax error
                else{
                    expectedSymbol = symbol;
                    return SYNTAXERROR;
                }
            }
            //Insert production
            symbolStack.push_back(-productionPos);
            addProduction(productionPos, symbolStack, true);
        }
    }
    //Extract info about the next reduction based on the next reduction symbol
    updateReductionInfo(-symbolStack.back());
    return GOOD;
}

//Determine LHS symbol (incremented), symbol count, and production number of a production given its position in the grammar
void LLParser::updateReductionInfo(int prodPos){
    curSymbolCount = grammar[prodPos];
    for (int i=0; i<ruleNumStart.size(); i++){
        //Find rule whose starting position exceeds production position and recognize its lhs symbol
        if (ruleStart(i) > prodPos) {
            curLhs = toRuleNum(i-1);
            break;
        }
    }
    //Loop through each production of the rule until the right one is found
    curProdNum = 0;
    for (int j=ruleStart(toRuleCount(curLhs)); j<prodPos; j=nextProduction(j)){
        curProdNum++;
    }
}

//Adds ParseValue of current token, which is the token string on the heap
void LLParser::addTokenValue(){
    std::string *stringPtr = new std::string();
    for (char* c=prevpos; c<curpos; c++){
        stringPtr->push_back(*c);
    }
    ParseValue value;
    value.ptr = stringPtr;
    value.toDelete = true;
    valueStack.push_back(value);
}